### PR TITLE
Add new method to validate usernames

### PR DESCRIPTION
**Pull request by mevan-karu**
### Proposed changes in this pull request
Changes to SelfRegistrationMgtClient to support username validation in the backend.
https://github.com/wso2/product-is/issues/8117

### CODE REVIEW

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/SelfRegistrationMgtClient.java [hunk 1]
```diff
@@ -210,60 +211,85 @@ public Integer checkUsernameValidity(String username) throws SelfRegistrationMgt
...
             JSONObject property = new JSONObject();
             property.put(
                     IdentityManagementEndpointConstants.KEY, IdentityManagementEndpointConstants.SKIP_SIGN_UP_ENABLE_CHECK);
             property.put(IdentityManagementEndpointConstants.VALUE, skipSignUpCheck);
             properties.put(property);
-            user.put(PROPERTIES, properties);
+
+            if (!StringUtils.isBlank(user.getTenantDomain())) {
+                JSONObject tenantProperty = new JSONObject();
+                tenantProperty.put(IdentityManagementEndpointConstants.KEY, IdentityManagementEndpointConstants.TENANT_DOMAIN);
```
Comment (thanujalk) [inline components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/SelfRegistrationMgtClient.java:249]: exceeds 120 limit.

Comment (mevan-karu) [reply components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/SelfRegistrationMgtClient.java]: Fixed

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/SelfRegistrationMgtClient.java [hunk 2]
```diff
@@ -210,60 +211,86 @@ public Integer checkUsernameValidity(String username) throws SelfRegistrationMgt
     public Integer checkUsernameValidity(String username, boolean skipSignUpCheck) throws
             SelfRegistrationMgtClientException {
 
+        User user = new User();
+        user.setUsername(username);
```
Comment (emswbandara) [inline components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/SelfRegistrationMgtClient.java:215]: don't we need to set tenantDomain to the user object?

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/SelfRegistrationMgtClient.java [hunk 3]
```diff
@@ -201,7 +202,7 @@ public Integer checkUsernameValidity(String username) throws SelfRegistrationMgt
     /**
      * Checks whether a given username is valid or not.
      *
-     * @param username Username.
+     * @param username
```
Comment (emswbandara) [inline components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/SelfRegistrationMgtClient.java:205]: do we need this change?

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/SelfRegistrationMgtClient.java [hunk 4]
```diff
@@ -210,60 +211,86 @@ public Integer checkUsernameValidity(String username) throws SelfRegistrationMgt
...
 
             JSONArray properties = new JSONArray();
             JSONObject property = new JSONObject();
             property.put(
                     IdentityManagementEndpointConstants.KEY, IdentityManagementEndpointConstants.SKIP_SIGN_UP_ENABLE_CHECK);
             property.put(IdentityManagementEndpointConstants.VALUE, skipSignUpCheck);
             properties.put(property);
-            user.put(PROPERTIES, properties);
+
+            if (!StringUtils.isBlank(user.getTenantDomain())) {
```
Comment (emswbandara) [inline components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/SelfRegistrationMgtClient.java:247]: let's use StringUtils.isNotBlank here

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/SelfRegistrationMgtClient.java (hunk 2 above)
Comment (mevan-karu) [reply components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/SelfRegistrationMgtClient.java:215]: This was the method used for preprocessed usernames. We can deprecate this method.

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/SelfRegistrationMgtClient.java (hunk 4 above)
Comment (mevan-karu) [reply components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/SelfRegistrationMgtClient.java:247]: Fixed

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/SelfRegistrationMgtClient.java (hunk 2 above)
Comment (emswbandara) [reply components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/SelfRegistrationMgtClient.java:215]: let's deprecate.. add the deprecate annotation and add a comment to use new method

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/SelfRegistrationMgtClient.java [hunk 5]
```diff
@@ -206,64 +207,92 @@ public Integer checkUsernameValidity(String username) throws SelfRegistrationMgt
      *                        tenant.
      * @return An integer with status code.
      * @throws SelfRegistrationMgtClientException Self Registration Management Exception.
+     * @Deprecated Use {@link #checkUsernameValidity(User user, boolean skipSignUpCheck)}
```
[the remark this answers is not in the document]
Comment (mevan-karu) [reply components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/SelfRegistrationMgtClient.java:210]: Fixed in https://github.com/wso2/carbon-identity-framework/pull/3035

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/SelfRegistrationMgtClient.java [hunk 6]
```diff
@@ -206,64 +207,92 @@ public Integer checkUsernameValidity(String username) throws SelfRegistrationMgt
...
+     * @return An integer with status code.
+     * @throws SelfRegistrationMgtClientException Self Registration Management Exception.
+     */
+    public Integer checkUsernameValidity(User user, boolean skipSignUpCheck) throws
+            SelfRegistrationMgtClientException {
+
         if (log.isDebugEnabled()) {
-            log.debug("Checking username validating for username: " + username + ". SkipSignUpCheck flag is set to "
-                    + skipSignUpCheck + ".");
+            log.debug("Checking username validating for username: " + user.getUsername() +
```
[the remark this answers is not in the document]
Comment (mevan-karu) [reply components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/SelfRegistrationMgtClient.java:234]: Fixed in https://github.com/wso2/carbon-identity-framework/pull/3035

**Pull request merged**